### PR TITLE
Highlight identified point cloud points in 3d views

### PR DESCRIPTION
**Pull request by uclaros**
## Description
Similar to how points are highlighted on the 2d map canvas, identified points now also get highlighted on 3d views using a QgsRubberBand3D, which is now extended to support multi-points too.
image

The highlights are rendered on top of the 3d scene and are not occluded by the scene's contents

Peek 2024-12-06 17-51

and are scaled to follow the 3d renderer's point size

image

### CODE REVIEW

File: src/3d/qgs3dmapcanvas.cpp [hunk 1]
```diff
@@ -13,6 +13,9 @@
  *                                                                         *
  ***************************************************************************/
 
+#include "qgsframegraph.h"
+#include "qgspointcloudlayer3drenderer.h"
+#include "qgsrubberband3d.h"
```
Comment (wonder-sk) [inline src/3d/qgs3dmapcanvas.cpp:18]: can we have these includes together with other qgis includes a bit further below?

File: src/3d/qgs3dmapcanvas.cpp [hunk 2]
```diff
@@ -348,3 +351,60 @@ QVector<QgsPointXY> Qgs3DMapCanvas::viewFrustum2DExtent()
...
+
+void Qgs3DMapCanvas::highlightFeature( const QgsFeature &feature, QgsMapLayer *layer )
+{
+  // we only support point clouds for now
+  if ( layer->type() != Qgis::LayerType::PointCloud )
+    return;
+
+  const QgsGeometry geom = feature.geometry();
+  QgsPoint pt( geom.vertexAt( 0 ) );
+  pt.setZ( pt.z() / mMapSettings->terrainVerticalScale() );
```
Comment (wonder-sk) [inline src/3d/qgs3dmapcanvas.cpp:363]: hmm this is a bit tricky - point's Z value may not depend on terrain's vertical scale (and point clouds are never clamped to terrain).  I would skip this Z update, and then for vector layers it can be adjusted based on how the Z value gets calculated...

File: src/3d/qgs3dmapcanvas.h [hunk 3]
```diff
@@ -240,6 +255,9 @@ class _3D_EXPORT Qgs3DMapCanvas : public QWindow
     QString mCaptureFileFormat;
 
     QgsTemporalController *mTemporalController = nullptr;
+
+    //! This holds rubber bands for highlighting identified features
+    QMap<QgsMapLayer *, QgsRubberBand3D *> mHighlights;
```
Comment (wonder-sk) [inline src/3d/qgs3dmapcanvas.h:260]: may be good to add info about ownership of pointers.

shall we clear highlights in the destructor?

File: src/3d/qgsrubberband3d.cpp [hunk 4]
```diff
@@ -187,7 +228,6 @@ void QgsRubberBand3D::updateGeometry()
 
 void QgsRubberBand3D::updateMarkerMaterial()
 {
-  delete mMarkerMaterial;
```
Comment (wonder-sk) [inline src/3d/qgsrubberband3d.cpp:190]: why not delete the marker material anymore?

File: src/3d/qgsrubberband3d.cpp [hunk 5]
```diff
@@ -88,58 +91,93 @@ QgsRubberBand3D::QgsRubberBand3D( Qgs3DMapSettings &map, QgsWindow3DEngine *engi
...
   {
     static_cast<QgsMarkerSymbolLayer *>( mMarkerSymbol->symbolLayer( 0 ) )->setStrokeColor( color );
   }
   updateMarkerMaterial();
 }
 
+void QgsRubberBand3D::setMarkerType( MarkerType marker )
+{
+  if ( mMarkerSymbol )
+    delete mMarkerSymbol;
```
Comment (wonder-sk) [inline src/3d/qgsrubberband3d.cpp:155]: a nitpick - no need to test `if ( mMarkerSymbol )` - we can just delete it even when null. I would also move the deletion just above the creation of new marker symbol. (Or we could even switch to unique_ptr for less hassle?)

File: src/3d/qgs3dmapcanvas.cpp [hunk 6]
```diff
@@ -348,3 +351,60 @@ QVector<QgsPointXY> Qgs3DMapCanvas::viewFrustum2DExtent()
...
+
+  if ( !mHighlights.contains( layer ) )
+  {
+    QgsRubberBand3D *band = new QgsRubberBand3D( *mMapSettings,
+        mEngine,
+        mEngine->frameGraph()->rubberBandsRootEntity(),
+        Qgis::GeometryType::Point );
+    band->setHideLastMarker( false );
```
Comment (wonder-sk) [inline src/3d/qgs3dmapcanvas.cpp:371]: if showing last marker is the default, we can get rid of this line?

File: src/app/qgsidentifyresultsdialog.cpp [hunk 7]
```diff
@@ -1395,7 +1395,7 @@ void QgsIdentifyResultsDialog::addFeature( QgsPointCloudLayer *layer, const QStr
     feature.setAttribute( fieldIndex, v );
   }
 
-  const QgsGeometry selectionGeometry( QgsGeometry::fromPointXY( QgsPointXY( attributes[QStringLiteral( "X" )].toDouble(), attributes[QStringLiteral( "Y" )].toDouble() ) ) );
+  const QgsGeometry selectionGeometry( QgsGeometry::fromPoint( QgsPoint( attributes[QStringLiteral( "X" )].toDouble(), attributes[QStringLiteral( "Y" )].toDouble(), attributes[QStringLiteral( "Z" )].toDouble() ) ) );
```
Comment (wonder-sk) [inline src/app/qgsidentifyresultsdialog.cpp:1398]: Does this work when offset/scale is applied to point cloud in layer properties? In such case, the "Z" attribute should be replaced by "Z (original)" and "Z (adjusted)" (translatable strings)

File: src/3d/qgs3dmapcanvas.cpp (hunk 2 above)
Comment (uclaros) [reply src/3d/qgs3dmapcanvas.cpp:363]: This is weird, I copied this logic from the 3d measure tool and they both work as expected.

It seems there's a bug in `Qgs3DUtils::clampAltitude` and `Qgs3DUtils::clampAltitudes` functions:
`AltitudeClamping::Absolute` is treated the same as `AltitudeClamping::Relative` so terrainZ and vertical scale are always applied.

I'll skip the Z update here but it will not render correctly when terrain vertical scale is set, until we fix the utils functions.

File: src/3d/qgs3dmapcanvas.h (hunk 3 above)
Comment (uclaros) [reply src/3d/qgs3dmapcanvas.h:260]: yes, indeed

File: src/3d/qgsrubberband3d.cpp (hunk 4 above)
Comment (uclaros) [reply src/3d/qgsrubberband3d.cpp:190]: `mMarkerMaterial` is added as a component to `mMarkerEntity`.
The docs say that is component has no parent, ownership is transferred, so we shouldn't be deleting it

File: src/3d/qgsrubberband3d.cpp (hunk 5 above)
Comment (uclaros) [reply src/3d/qgsrubberband3d.cpp:155]: switched to unique pointer

File: src/app/qgsidentifyresultsdialog.cpp (hunk 7 above)
Comment (uclaros) [reply src/app/qgsidentifyresultsdialog.cpp:1398]: Ah, good catch, it's a bit ugly to rely on a translatable string here...
I fixed this, but can't think of an easy way to update existing rubberbands when the layer's scale/offset changes.

**Review by wonder-sk — APPROVED** [review]
